File: 113-path-sum-ii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        result = list()
        cur_path = list()
        self._pathSum(root, cur_path, sum, result)
        return result
        
    def _pathSum(self, root, cur_path, sum, result):
        if not root: return
        
        cur_path.append(root.val)
        if root.val == sum and not root.left and not root.right:
            result.append(cur_path[:])
        
        self._pathSum(root.left, cur_path, sum - root.val, result)
        self._pathSum(root.right, cur_path, sum - root.val, result)
        cur_path.pop()
        

File: 113-path-sum-ii/path-sum-ii.go
/*
 * @lc app=leetcode id=113 lang=golang
 *
 * [113] Path Sum II
 *
 * https://leetcode.com/problems/path-sum-ii/description/
 *
 * algorithms
 * Medium (40.18%)
 * Total Accepted:    224.8K
 * Total Submissions: 559K
 * Testcase Example:  '[5,4,8,11,null,13,4,7,2,null,null,5,1]\n22'
 *
 * Given a binary tree and a sum, find all root-to-leaf paths where each path's
 * sum equals the given sum.
 *
 * Note: A leaf is a node with no children.
 *
 * Example:
 *
 * Given the below binary tree and sum = 22,
 *
 *
 * ⁠     5
 * ⁠    / \
 * ⁠   4   8
 * ⁠  /   / \
 * ⁠ 11  13  4
 * ⁠/  \    / \
 * 7    2  5   1
 *
 *
 * Return:
 *
 *
 * [
 * ⁠  [5,4,11,2],
 * ⁠  [5,8,4,5]
 * ]
 *
 *
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSumII(root *TreeNode, sum int, res [][]int, current []int) [][]int {
	if root == nil {
		return res
	}

	current = append(current, root.Val)
	if root.Val == sum && root.Left == nil && root.Right == nil {
		res = append(res, append([]int(nil), current...))
		current = current[:len(current)-1]
	}
	res = pathSumII(root.Left, sum-root.Val, res, current)
	res = pathSumII(root.Right, sum-root.Val, res, current)
	return res
}
func pathSum(root *TreeNode, sum int) [][]int {
	var res [][]int
	var current []int
	return pathSumII(root, sum, res, current)
}
